pyro/distributions/quantized.py

```diff
@@ -0,0 +1,256 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import math
+
+import torch
+from torch.distributions.utils import broadcast_all, lazy_property
+from torch.distributions import constraints
+
+from .conjugate import BetaBinomial
+from .torch import Beta, Binomial
+from .torch_distribution import TorchDistribution
+
+
+def quantize(rate):
+    """
+    Quantize a real-valued tensor by randomly rounding to the nearest two
+    integer neighbors, based on distance to each of those neighbors.
+    """
+    with torch.no_grad():
+        lb = rate.unsqueeze(-1).floor()
+        ub = lb + 1
+        counts = torch.cat([lb, ub], dim=-1)
+    probs = rate - lb
+    return probs, counts
+
+
+@torch.no_grad()
+def dequantize(count, int_dist):
+    """
+    Given a sample ``count`` from an nonnegative integer valued distribution
+    ``int_dist``, add dithering noise such that the distribution is preserved
+    under the :func:`quantize` method.
+
+    The following two models result in identical distributions over ``z``:
+
+        # Model 1.
+        z ~ int_dist
+
+        # Model 2.
+        x ~ int_dist
+        y = dequantize(x, int_dist)
+        probs, counts = quantize(y)
+        b ~ Bernoulli(probs)
+        z = b + counts[..., 0]
+    """
+    neighbor = count.new_full(count.shape, 0.5)
+    neighbor.bernoulli_().mul_(2).add_(-1).add_(count).abs_()
```

Comment:
I think we should use `uniform_()` instead of `bernoulli_()` here; but even so, the math here is not quite convincing. I'll think more about this...

Reply:
I've rewritten this and factored it into the `DequantizedDistribution` class. I think the math is clearer now, but I still haven't written tests.

Reply:
Nice use of triangular distribution! I have tried to construct something similar (linear density on one side with predefined mean). I'll double check the math.